### ISSUE TITLE
Proper SFTP Support and Framework for Alternative Remote File System Types

### ISSUE THREAD

**Issue by jnterry** [jnterry/treeploy#9]
For context, previous idea for deployment to remote systems was to use sshfs, that was deemed infeasible, see issue: https://github.com/jnterry/treeploy/issues/3

An alternative would be to actually connect to the remote server using SFTP (a subsystem of SSH allowing for remote file transfer). This would not change the requirements of the remote sever (SSHFS relies on SFTP being enabled) - but would mean that the local server running treeploy would not need to have sshfs installed.

My suggestion would be to introduce a new "FileDriver" interface which has minimum set of methods we need in order to do a deployment (make directory, read file, stat file, etc) and then implement a driver for the local filesystem and another for remote file systems. 

The code would need to be refactored to create a driver instance for the source directory and target directory, and then make all file system calls through these - hence what the file system is on each end is abstracted from the main loop of deployment.

This pattern would additionally allow further file drivers to be added in the future, for example maybe some "read only" file drivers that can only be used for the source directory, to allow deploying from a http webserver (with directory listings), or from an amazon S3 bucket, etc.

A few changes need to be made before this is possible:
- We can't have a "copy file" method as we need to go from source driver to target driver. Instead have a getFileReadStream and getFileWriteStream methods, and then pipe one to the other (https://stackoverflow.com/a/11295106)
- No synchronous libraries exist for SFTP/SSH -> the file drivers would need to expose a promise based interface, thus existing code needs to be moved away from using synchronous calls

Suggested path to implementing this:
- Convert all synchronous code to promise based code (issue https://github.com/jnterry/treeploy/issues/6)
- Create a local FileDriver that just proxies methods in fs or fs-extra, test everything still works
- Create a remote FileDriver that proxies requests through SFTP

----------------------

With regards to implementing the remote FileDriver there are some issues. Ideally we want root access on the remote to change permissions, etc - however we should support ssh connections as non-root user as we may not want to configure remote server to allow root login over ssh.
This would be fine, except that SFTP does not have any concept of privlledge escalation (eg, sudo, su, etc).

A library like "ssh2" does not support this sort of thing, see: https://github.com/mscdex/ssh2/issues/287

Suggested workaround is to use make the remote FileDriver utalize standard shell commands with sudo, using cat and tee to make file streams: https://github.com/mscdex/ssh2/issues/626

Below is proof of concept code for connecting and making a directory:
```yaml
var Client = require('ssh2').Client;

var conn = new Client();

conn.on('ready', function() {
  console.log('Client :: ready');
  conn.exec('sudo mkdir /test-ssh', function(err, stream) {
    if (err) throw err;
    stream.on('close', function(code, signal) {
      console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
      conn.end();
    }).on('data', function(data) {
      console.log('STDOUT: ' + data);
    }).stderr.on('data', function(data) {
      console.log('STDERR: ' + data);
    });
  });
}).connect({
  host: 'home-swarm-01',
  port: 22,
  username: 'ansible',
  agent: '/tmp/ssh-Ce9FbHFxsyBY/agent.6032',
});
```

Note that the agent field of connect needs to be changed to get the actual unix socket of the running ssh agent, since it changes each time.

----------- 

In summary, advantages of new method:
- No need for sshfs on machine running treeploy
- Framework in place for future support of other remote file systems (http server, amazon s3 bucket, etc)

However its probably at least 15-20 hours of work (5 to convert current sync code to promises, 5 to make local file driver, 5-10 to make remote file driver, then another however many to solve whatever crops up).

The CLI and API interfaces would not need to be changed in an incompatible way to implement this change - hence this will be delayed until after the 1.0.0 release.

**Comment by jnterry** [comment]
Implemented and merged into master in: 823b229c53badeee8ebea7248dde9e1606a1a5a2

Released as v1.0.0
